### DISCUSSION TITLE
Guidelines for design thinking

### DISCUSSION THREAD

**Discussion by WildWeazel** [C7-Game/Prototype#243] [Ideas]
These are some ideas I've come up when thinking about various design issues in the abstract. The overall theme is to decouple the way our code is structured from the way (we assume) Civ3 works. They're more guidelines than hard and fast rules, but are worth considering whenever you're writing a class, interface, or mechanic.

* Divide the game mechanics into subsystems, and [use events to communicate](https://refactoring.guru/design-patterns/observer) between subsystems as much as possible and internally when it makes sense. Subsystems should be highly [cohesive](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) sets of mechanics that likely produce all of some group of related events: unit combat, or city economy, or culture. This is mainly represented by the "component-event model" that I'm taking charge of, and will start fleshing out next. See also: [Single responsibility](https://en.wikipedia.org/wiki/Single-responsibility_principle)
* All game event types (as opposed to UI interaction eg. button clicks) should be declared in a single namespace and named and parameterized for a specific context: ie, a single source of all game event definitions with no reusing the same types for different purposes. There should be one reason for a given event to be raised, and that reason should always raise the same event. We _may_ even implement a central event bus (ie the [publish-subscribe pattern](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern)) for all events to be visible to all components.
* For complex sequences of events, a subsystem should expose a general event with a contextual object that exposes more detailed events. Example: the unit combat system exposes onUnitCombat(Battle), while Battle has onDamage, onDeath, onVictory, etc. Other components can just listen for onUnitCombat and then decide if they want to subscribe to its other events.
* If it uses a value, make it configurable. If it's configurable, consider making it context specific. In other words, avoid hard-coded values wherever possible. Not just cases where Civ3 has a cap, but anywhere that a specific number is used in a decision or algorithm. Even try to avoid the implicit "one", if it may make sense to allow a set of things. Examples: obviously railroad speed should be configurable, but what if the modder wants to make it terrain dependent or increase it with a tech, how would you code for that? Tiles can have (rail)road + another improvement, but what if there was an arbitrary set of improvements?
* Code rules as generically as possible such that they can be easily configured down to the default case. Example: resource yields and consumption. In Civ3 every resource is worth 1 if roaded, and only trade deals consume it. But if a resource provides X units, an improvement A yields Y units, and things requiring the resource consume Z units, that allows for the Civ3 behavior just by configuring all cases to X=1, A = road, Y = 1, Z = 0. But also provides much more flexibility.
* Similarly, hold references to [interfaces](https://en.wikipedia.org/wiki/Dependency_inversion_principle), and generic collections of interfaces, wherever possible. Example: Instead of Tax, Science, and Luxuries, the budget could have a set of "IBudgetItem" which by default would be those 3 but could accept any number of others.
* Support modular additions: in core components, scripts, or mod content. Look for places where new functionality could be plugged in, either by responding to an event or providing an object/callback like the above example. Allow for data to come from different sources both static and dynamic. Consider DLC-like mods that add a new civ with a unique unit, event script, etc to an existing mod structure.

And here are some specific design patterns that will likely be useful for certain things:

* [State pattern](https://refactoring.guru/design-patterns/state) for high-level handling: delegate the Process and other Godot lifecycle calls to a handler specific to the active state. Handle input, drawing, etc only in the appropriate method for the current state. We kind of have the outline of this already in Game.cs.
* [Prototype pattern](https://refactoring.guru/design-patterns/prototype) for things like units that are instantiated many times from a specific base type.
* [Repository pattern](https://martinfowler.com/eaaCatalog/repository.html) to encapsulate reading and writing file data without having to know about BIQ, SAV, JSON or other formats. There will likely end up being a Data Abstraction Layer between the core and the Civ3/C7/etc file IO to enforce this.
* [Command pattern](https://refactoring.guru/design-patterns/command) for user actions sent to the core, or unit abilities (buttons) reported to the UI
* [Facade pattern](https://refactoring.guru/design-patterns/facade) and/or [small interfaces](https://en.wikipedia.org/wiki/Interface_segregation_principle) exposing actionable features of the core to the UI
* [Strategy pattern](https://refactoring.guru/design-patterns/strategy) for behaviors that could be swapped out or provided by a script, even big stuff like AI personalities and map generation.
* [Builder pattern](https://refactoring.guru/design-patterns/builder) for when you need to specify complex things step by step, like the new game parameters or a production queue. 
* [Flyweight pattern](https://refactoring.guru/design-patterns/flyweight) for optimizing big collections of things like tiles or units that share most data.

In short, read up on [SOLID principles](https://en.wikipedia.org/wiki/SOLID), [event-driven architecture](https://en.wikipedia.org/wiki/Event-driven_architecture), [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection), and [OO design patterns](https://refactoring.guru/design-patterns), and try to design a mechanic that _can_ work like Civ3, but doesn't have to if you can think of an alternative.

These are just some starting points for thinking. Question and discuss. If you come up with more, let's collect them for future reference.
